cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BaseRDDCollection.java

```diff
@@ -315,6 +317,10 @@ public void publishAlerts(StageSpec stageSpec, StageStatisticsCollector collecto
     return new RDDCollection<>(sec, functionCacheFactory, jsc, sqlContext, datasetContext, sinkFactory, rdd);
   }
 
+  protected <U> RDDCollection<U> wrap(BatchSQLEngineAdapter adapter, SQLEngineJob<SQLDataset<U>> job) {
```

Comment:
Ok seems like this is the reason the rdd might be null? I don't think it's a good idea to modify RDDCollection such that it can be backed by an RDD or some data in an external SQL system.  Instead, it seems better to have a separate "SparkCollection" class backed just by the SQLJob. 

However, I'm also not sure we need to fit it in the SparkCollection abstraction at all yet, since you can't do anything but join right now. See comments in the spark driver.

Reply:
Yes, the RDD can be null if this RDDCollection represents a dataset that exists in the SQL engine.

In this case, we pull only if a method needs to operate on the RDD. Since joins run on the SQL engine, we will only pull data from the SQL engine if another (non-join) operation needs to operate on the RDD.

Reply:
I have removed this logic from the RDD Collection and implemented the `SQLEngineCollection` for SQL engine backed collections.